src/blockdata.rs

```diff
@@ -125,23 +133,73 @@ impl Block {
-        let len = self.get_aggregated_public_key_length();
-        let bytes = &self.payload()[Self::XFIELD_POSITION..Self::XFIELD_POSITION + len];
-        PublicKey::from_slice(bytes).ok()
+        match self.0[Self::XFIELD_POSITION - 1] {
+            1 => {
+                let len = self.get_xfield_length().expect("can not get xfield size");
+                let bytes = &self.payload()[Self::XFIELD_POSITION + len.len()
+                    ..Self::XFIELD_POSITION + len.len() + len.0 as usize];
+                PublicKey::from_slice(bytes).ok()
+            }
+            _ => None,
+        }
     }
 
-    /// the length of aggregated public key.
+    /// the length of xfield.
+    /// return 0  if xfieldType is None.
     /// return 33 if xfieldType is AggregatePublicKey.
-    /// return 0 otherwise
-    fn get_aggregated_public_key_length(&self) -> usize {
+    /// return value of the first byte if xfieldType is unknown,
+    /// this assumes that the leading of the new xfieldType added in the future is the length of the xfield value.
+    fn get_xfield_length(&self) -> Option<VarInt> {
         match self.0[Self::XFIELD_POSITION - 1] {
-            1 => 33,
-            _ => 0,
+            0 => None,
+            _ => {
+                let n = self.0[Self::XFIELD_POSITION];
+                match n {
+                    0xFF => {
+                        let mut slice = [0u8; 8];
+                        slice.copy_from_slice(
+                            &self.0[Self::XFIELD_POSITION + 1..Self::XFIELD_POSITION + 9],
+                        );
+                        let x = u64::from_le_bytes(slice);
+                        if x < 0x100000000 {
+                            None
+                        } else {
+                            Some(VarInt(x))
+                        }
+                    }
+                    0xFE => {
+                        let mut slice = [0u8; 4];
+                        slice.copy_from_slice(
+                            &self.0[Self::XFIELD_POSITION + 1..Self::XFIELD_POSITION + 5],
+                        );
+                        let x = u32::from_le_bytes(slice);
+                        if x < 0x10000 {
+                            None
+                        } else {
+                            Some(VarInt(x as u64))
+                        }
+                    }
+                    0xFD => {
+                        let mut slice = [0u8; 2];
+                        slice.copy_from_slice(
+                            &self.0[Self::XFIELD_POSITION + 1..Self::XFIELD_POSITION + 3],
+                        );
+                        let x = u16::from_le_bytes(slice);
+                        if x < 0xFD {
+                            None
+                        } else {
+                            Some(VarInt(x as u64))
+                        }
+                    }
+                    n => Some(VarInt(n as u64)),
+                }
```

Comment:
How about using decoding function for VarInt like bellow? 

Another point, I think it should be some error when parsing VarInt is a failure. Now it returns None.

```suggestion
               let mut slice = &self.0[Self::XFIELD_POSITION..];
                match VarInt::consensus_decode(&mut slice) {
                    Ok(i) => Some(i),
                    Err(_) => // should error.
                }
```

Reply:
fixed at 50cae2d
and I added validation for xfield when 
  - master gets the block via getnewblock rpc.
  - member gets the block via candidateblock message.